### ISSUE TITLE
Question about data loading during training

### ISSUE THREAD

**Issue by Petros626** [ultralytics/ultralytics#19136]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I want to know, if the samples from the training folder are loaded randomly during the training of following the folder order (1.png, 2.png, 3.png etc.).

Thanks

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Petros626, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

🔎 An Ultralytics engineer will review your question soon and provide assistance!

**Comment by Petros626** [comment]
can you specify part of the code?

**Comment by glenn-jocher** [comment]
@Petros626 **Answer:** Ultralytics YOLO training loads images in **random order** by default when using standard training (non-rectangular). The shuffle is implemented in the data loader to ensure proper generalization. For rectangular training (`rect=True`), images are sorted by aspect ratio for efficiency but still sampled randomly in batches. You can review the data loading implementation in our [BaseDataset class](https://docs.ultralytics.com/reference/data/base/#ultralytics.data.base.BaseDataset). 

Let me know if you need clarification on specific implementation details!

**Comment by Petros626** [comment]
@glenn-jocher  Thank you, I just wanted to see the specific line in the code.

**Comment by glenn-jocher** [comment]
For non-rectangular training (`rect=False`), image loading order randomization occurs in the DataLoader's shuffle parameter here:  
https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/build.py#L157  
https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L55  

The actual image path list is initially gathered sorted via `get_img_files()`, then shuffled by the DataLoader when `rect=False`. Let me know if you'd like me to clarify any specific implementation details!

**Comment by glenn-jocher** [comment]
You're welcome! Feel free to reach out if you have any other questions. 😊
